### PR TITLE
Add offsets-based reduction to segment_reduce (CPU, CUDA)

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Introduce an `offsets` kwarg to `segment_reduce` to support reductions for sparse tensors in the CSR format. 

Offsets are related to the existing `lengths` kwarg via the following:

```
zeros_shape = list(lengths.shape)
zeros_shape[-1] = 1
offsets = torch.cat((torch.zeros(zeros_shape), lengths}, -1)
offsets.cumsum_(-1)
```

And `lengths` can be obtained from `offsets` via `torch.diff(offsets)`

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78908
* #78918
* __->__ #78907
* #77061

### CODE REVIEW

File: aten/src/ATen/native/cuda/SegmentReduce.cu [hunk 1]
```diff
@@ -372,27 +382,59 @@ Tensor _segment_reduce_cuda_backward_kernel(
...
+    grad_contig, output_contig, data_contig, reduction, offsets_contig, axis, initial, /*is_offsets_like=*/true);
+}
+
+Tensor _segment_reduce_lengths_offsets_cuda_kernel(
+  SegmentReductionType reduction,
+  const Tensor& data,
+  const Tensor& lengths_or_offsets,
+  int64_t axis,
+  const c10::optional<Scalar>& initial,
+  bool is_offsets_like) {
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/cuda/SegmentReduce.cu:415]: Tried to make `is_offsets_like` a template variable like `is_scatter_like` in ScatterGatherKernel, but oddly enough that caused errors like `<data> : undeclared identifier` (and similarly for all the other variables in the nested lambda) specifically on Windows. I've noticed that windows has problems with capturing variables in nested lambdas from [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L860-L861). But in this case removing the template and passing `is_offsets_like` (and changing absolutely nothing else the lambda is still nested) seemed to fix it but I'm not sure why this is the case. Do you have any thoughts on how else I could fix this?

**Review by cpuhrsch — APPROVED** [review]
Great! Thank you!

**Comment by osalpekar** [comment]
This PR appears to have broken PyTorch Distributed and fx2trt builds internally. See the error below. More can be found in the Jobs tab of the corresponding diff: D37114159 

```
Summary: 
stderr: buck-out/dev/gen/caffe2/generate-code/torch/csrc/autograd/generated/VariableType_2.cpp:6610:78: error: no viable conversion from 'c10::SymIntArrayRef' to 'at::IntArrayRef' (aka 'ArrayRef<long>')
    at::redispatch::expand_copy_outf(ks & c10::after_autograd_keyset, self_, size, implicit, out_);
                                                                             ^~~~
stderr: buck-out/dev/gen/caffe2/generate-code/torch/csrc/autograd/generated/VariableType_2.cpp:6610:78: error: no viable conversion from 'c10::SymIntArrayRef' to 'at::IntArrayRef' (aka 'ArrayRef<long>')
    at::redispatch::expand_copy_outf(ks & c10::after_autograd_keyset, self_, size, implicit, out_);
                                                                             ^~~~
```

```
Summary: 
stderr: buck-out/dev/gen/caffe2/generate-code/torch/csrc/autograd/generated/python_torch_functions_1.cpp:4495:45: error: no viable conversion from 'c10::SymInt' to 'int64_t' (aka 'long')
        return self.narrow_copy(dim, start, length);
                                            ^~~~~~
stderr: buck-out/dev/gen/caffe2/generate-code/torch/csrc/autograd/generated/python_torch_functions_1.cpp:4495:45: error: no viable conversion from 'c10::SymInt' to 'int64_t' (aka 'long')
        return self.narrow_copy(dim, start, length);
                                            ^~~~~~
```

**Comment by mikaylagawarecki** [comment]
@osalpekar I'm not too sure whether this diff caused those failures, I reran the PT Distributed and fx2trt builds on D37114159 and they are passing now. Is it ok to reland this PR?

**Pull request closed without merging**